File: site/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/pfi/ProcessingOrdersTest.kt
package website.tbd.developer.site.docs.tbdex.pfi

import tbdex.sdk.protocol.models.Order
import tbdex.sdk.protocol.models.OrderStatus
import tbdex.sdk.protocol.models.OrderStatusData
import tbdex.sdk.protocol.models.Close
import tbdex.sdk.protocol.models.CloseData
import website.tbd.developer.site.docs.utils.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.*

class ProcessingOrdersTest {

    private lateinit var dataProvider: MockDataProvider
    private val pfiDid = TestData.PFI_DID
    private val customerDid = TestData.ALICE_DID
    private lateinit var exchangesApiProvider: ExchangesApiProvider
    private lateinit var orderMessage: Order

    @BeforeEach
    fun setup() {
        dataProvider = MockDataProvider()
        exchangesApiProvider = ExchangesApiProvider()

        orderMessage = TestData.getOrder(
            to = pfiDid.uri,
            from = customerDid.uri
        )
    }

  @Test
  fun `PFI Accesses Private Data`() {
    val rfq = TestData.getRfq()

    // :snippet-start: pfiAccessPrivateDataKt
    val creditCardNumber = rfq.privateData!!.payin!!.paymentDetails!!["cardNumber"]
    // :snippet-end:

    assertNotNull(creditCardNumber)
  }
    @Test
    fun `PFI creates orderStatus and verifies it`() {
        exchangesApiProvider.setWrite()

        // :snippet-start: pfiOrderStatusKt
        val orderStatus = OrderStatus.create(
            from = pfiDid.uri,
            to = orderMessage.metadata.from,
            exchangeId = orderMessage.metadata.exchangeId,
            protocol = "1.0",
            orderStatusData = OrderStatusData("PROCESSING")
        )

        orderStatus.sign(pfiDid)
        exchangesApiProvider.write(orderStatus)
        // :snippet-end:

        assertEquals(orderStatus.data.orderStatus, "PROCESSING", "OrderStatus was not correctly set")
        assertNotNull(orderStatus.verify(), "OrderStatus signature is invalid")
    }


    @Test
    fun `PFI creates Close and verifies it`() {
        exchangesApiProvider.setWrite()

        // :snippet-start: pfiCloseOrderKt
        val closeMessage = Close.create(
            to = orderMessage.metadata.from,
            from = pfiDid.uri,
            exchangeId = orderMessage.metadata.exchangeId,
            protocol = "1.0",
            closeData = CloseData(
                reason = "COMPLETED",
                success = true // Indicates the transaction was successful
            )
        )

        closeMessage.sign(pfiDid)
        exchangesApiProvider.write(closeMessage)
        // :snippet-end:

        assertEquals(closeMessage.data.reason, "COMPLETED", "Close reason was not correctly set")
        assertNotNull(closeMessage.verify(), "Close signature is invalid")
    }
}

File: site/testsuites/testsuite-javascript/__tests__/tbdex/pfi/processingOrders.test.js
import { Order, OrderStatus, Close  } from '@tbdex/http-server';
import { DevTools } from '@tbdex/http-client';
import { DidDht } from '@web5/dids';
import { MockDataProvider } from '../../utils/mockDataProvider';
import { test, expect, describe, beforeAll, assert } from 'vitest';

let pfiDid;
let senderDid;
let dataProvider;
let orderMessage;

describe('PFI: Orders', () => {
    beforeAll(async () => {
        // Set up providers and DID
        pfiDid = await DidDht.create({
            options:{
                publish: true,
                services: [{
                    id: 'pfi',
                    type: 'PFI',
                    serviceEndpoint: 'https://example.com/'
                }]
            }
        });

        senderDid = await DidDht.create({
            options: { publish: true }
        });

        dataProvider = new MockDataProvider();
        orderMessage = Order.create({
            metadata: {
                from: senderDid.uri,
                to: pfiDid.uri,
                exchangeId: "MyExchange"
            }
        });
    });

    test('PFI Accesses Private Data', async () => {
        const rfq = await DevTools.createRfq({
            sender: senderDid,
            receiver: pfiDid
        });
        
        // :snippet-start: pfiAccessPrivateDataJs
        const creditCardNumber = rfq.privateData.payin.paymentDetails.cardNumber
        // :snippet-end:

        expect(creditCardNumber).toBeDefined();
    });

    test('PFI Creates OrderStatus', async () => {
        // :snippet-start: pfiOrderStatusJs
        const orderStatus = OrderStatus.create({
            metadata: {
                from: pfiDid.uri,
                to: orderMessage.metadata.from,
                exchangeId: orderMessage.metadata.exchangeId
            },
            data: { orderStatus: 'PROCESSING' }
        })
        
        await orderStatus.sign(pfiDid)
        dataProvider.insert(orderStatus)
        // :snippet-end:

        expect.soft(orderStatus.data.orderStatus).toBe('PROCESSING');

        try {
            await orderStatus.verifySignature();
            await orderStatus.verify();
        } catch(e) {
            assert.fail(`Failed to verify offering requirements: : ${e.message}`);
        }
    });

    test('PFI Creates Close', async () => {
        // :snippet-start: pfiCloseOrderJs
        const closeMessage = Close.create({
            metadata: { 
                from: pfiDid.uri, 
                to: orderMessage.metadata.from, 
                exchangeId: orderMessage.metadata.exchangeId
            },
            data: { 
                reason: 'COMPLETED',
                success: true // Indicates the transaction was successful
            }
        })
        
        await closeMessage.sign(pfiDid)
        dataProvider.insert(closeMessage)
        // :snippet-end:

        expect.soft(closeMessage.data.reason).toBe('COMPLETED');

        try {
            await closeMessage.verifySignature();
            await closeMessage.verify();
        } catch(e) {
            assert.fail(`Failed to verify offering requirements: ${e.message}`);
        }
    });
});